pkg/util/kubernetes/leaderelection/leaderelection_engine.go

```diff
@@ -0,0 +1,122 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+// +build kubeapiserver
+
+package leaderelection
+
+import (
+	"encoding/json"
+	"time"
+
+	log "github.com/cihub/seelog"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ld "k8s.io/client-go/tools/leaderelection"
+	rl "k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/client-go/tools/record"
+)
+
+func (le *LeaderEngine) getCurrentLeader(electionId, namespace string) (string, *v1.Endpoints, error) {
+	endpoints, err := le.coreClient.Endpoints(namespace).Get(electionId, metav1.GetOptions{})
+	if err != nil {
+		return "", nil, err
+	}
+
+	val, found := endpoints.Annotations[rl.LeaderElectionRecordAnnotationKey]
+	if !found {
+		return "", endpoints, nil
+	}
+
+	electionRecord := rl.LeaderElectionRecord{}
+	if err := json.Unmarshal([]byte(val), &electionRecord); err != nil {
+		return "", nil, err
+	}
+	return electionRecord.HolderIdentity, endpoints, err
+}
+
+// newElection creates an election.
+// If `namespace`/`election` does not exist, it is created.
+// `id` is the id if this leader, should be unique.
+func (le *LeaderEngine) newElection(electionId, namespace string, ttl time.Duration) (*ld.LeaderElector, error) {
+	// We first want to check if the Endpoint the Leader Election is based on exists.
+	_, err := le.coreClient.Endpoints(namespace).Get(electionId, metav1.GetOptions{})
+
+	if err != nil {
+		if errors.IsNotFound(err) == false {
+			return nil, err
+		}
+		_, err = le.coreClient.Endpoints(namespace).Create(&v1.Endpoints{
```

Comment:
If we try creating the endpoint and get a 403, is the current error message explicit enough, or does it need special handling?
Not having the permission if the endpoint is already here, having the perm is not required I guess

Reply:
If the permission aren't set, we have this log:
```
2018-02-07 15:22:33 UTC | ERROR | (runner.go:276 in work) | Error running check kubernetes_apiserver: temporary failure in leaderElection, will retry later: User "system:serviceaccount:default:datadog-agent" cannot list endpoints in the namespace "default".: "clusterrole.rbac.authorization.k8s.io \"datadog-agent\" not found" (get endpoints)
```